packages/snaps-rpc-methods/src/permitted/requestPermissions.ts

```diff
@@ -0,0 +1,104 @@
+import type { JsonRpcEngineEndCallback } from '@metamask/json-rpc-engine';
+import type {
+  PermissionConstraint,
+  PermittedHandlerExport,
+  SubjectPermissions,
+} from '@metamask/permission-controller';
+import type {
+  RequestPermissionParams,
+  RequestPermissionsResult,
+} from '@metamask/snaps-sdk';
+import type { Snap, SnapPermissions } from '@metamask/snaps-utils';
+import { hasProperty } from '@metamask/utils';
+import type { JsonRpcRequest, PendingJsonRpcResponse } from '@metamask/utils';
+import deepEqual from 'fast-deep-equal';
+
+import type { MethodHooksObject } from '../utils';
+
+export type RequestPermissionsHooks = {
+  getSnap: () => Snap | undefined;
+  /**
+   * @returns Requested permission result.
+   */
+  requestPermissions: (
+    dynamicPermissions: RequestPermissionParams,
+  ) => Promise<SubjectPermissions<PermissionConstraint>>;
+};
+
+/**
+ * The `snap_requestPermissions` method implementation.
+ * Requests dynamic permissions at the runtime.
+ *
+ *
+ * @param request - The JSON-RPC request object.
+ * @param response - The JSON-RPC response object.
+ * @param _next - The `json-rpc-engine` "next" callback. Not used by this
+ * function.
+ * @param end - The `json-rpc-engine` "end" callback.
+ * @param hooks - The RPC method hooks.
+ * @param hooks.getSnap - Get Snap hook.
+ * @param hooks.requestPermissions - Request permission hook.
+ */
+async function requestPermissionsImplementation(
+  request: JsonRpcRequest<RequestPermissionParams>,
+  response: PendingJsonRpcResponse<RequestPermissionsResult>,
+  _next: unknown,
+  end: JsonRpcEngineEndCallback,
+  { getSnap, requestPermissions }: RequestPermissionsHooks,
+): Promise<void> {
+  const snap = getSnap();
+
+  const requestedPermissions = request.params as SnapPermissions;
+  const dynamicPermissions = snap?.manifest.dynamicPermissions;
+
+  if (!dynamicPermissions) {
+    throw new Error('This Snap have no dynamic permissions specified.');
+  }
+
+  const requestedPermissionsKeys = Object.keys(
+    requestedPermissions,
+  ) as (keyof typeof dynamicPermissions)[];
+
+  requestedPermissionsKeys.forEach((permissionKey) => {
+    if (!hasProperty(dynamicPermissions, permissionKey)) {
+      throw new Error(
+        `Requested dynamic permission (${String(
+          permissionKey,
+        )}) does not match the allowed dynamic permissions specified in the manifest.`,
+      );
+    }
+
+    if (
+      !deepEqual(
+        requestedPermissions[permissionKey],
+        dynamicPermissions[permissionKey],
+      )
+    ) {
+      throw new Error(
+        `Requested dynamic permission caveats do not match the allowed dynamic permissions caveats for "${String(
+          permissionKey,
+        )}".`,
+      );
+    }
+  });
+
+  response.result = await requestPermissions(
```

Comment:
As agreed on Slack, I would prefer keeping the same patterns as used before for other Snap RPC methods. 

In my opinnion this is btter because of the clarity, consistence, easy understanding, maintenance and more extensive possibilities for future improvements. Having aliases would slightly reduce the code duplication, but complicate middleware and add additional friction to it, plus it would introduce a new way for adding and maintaining Snap RPC methods which I think we should not do now, but try to keep things more consistent. 

I would prefer keeping it in the way its implemented. I made implementation very close to the `wallet_requestPermissions`. In any case processing of the permissions is required so the PermissionController understands it (more specifically: caveats), so `processSnapPermissions` is added for dynamic permission requests.